### PR TITLE
OSOE-559: Make pnpm retries more robust to prevent errors

### PR DESCRIPTION
**Pull request by 0liver**
[OSOE-559](https://lombiq.atlassian.net/browse/OSOE-559)

[OSOE-559]: https://lombiq.atlassian.net/browse/OSOE-559?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1 [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+Write-Output "$MessagePrefix Acquiring lock..."
+
+$timeoutMs = 20000
+$sleepMs = 100
+$maxTries = $timeoutMs / $sleepMs
+$stream = $null
+
+# Try to get an exclusive lock on a lock file to guarantee only one thread will execute the given $Command at any given time.
+for ($i = 0; $i -lt $maxTries -and $null -eq $stream; $i++)
```
Comment (dministro) [inline Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1]: Use meaningful names.
```suggestion
for ($currentTry = 0; $currentTry -lt $maxTries -and $null -eq $stream; $i++)
```

File: Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1 [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    catch [System.IO.IOException]
+    {
+        # These are expected to happen, keep going.
+        Start-Sleep -Milliseconds $sleepMs
+    }
+    catch
+    {
+        # When an unexpected exception is thrown, fail the script.
+        Write-Output $PSItem.Exception
+        exit 1
```
Comment (dministro) [inline Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1:40]:
```suggestion
        # The finally block runs if an exit keyword stops the script. See: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_try_catch_finally?view=powershell-7.3#freeing-resources-using-finally
        exit 1
```

**Review by dministro — CHANGES_REQUESTED** [review]
Please mention powershell in the pre-requirements in the docs (Windows, Linux).

File: Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1 (hunk 2 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1:40]: I don't think we need to add this comment here, as this is the normal behavior of a finally block and should be expected. Or do you disagree?

Comment (dministro) [reply Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1:40]: This is just a suggestion but was not clear to me, so this is why I suggested this comment.

Comment (0liver) [reply Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1:40]: I see. I would still argue that we shouldn't comment on built-in language features.

In this concrete case, the `finally` block exist for exactly the purpose of **always** running, whatever happens inside the the `try` and `catch` blocks.

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets [hunk 3]
```diff
@@ -156,6 +169,7 @@
     <ItemGroup>
       <NodeJsExtensionsTargetFiles
         Include="%(NodeJsExtensionsTargetDirectories.Identity)/**/*"
+        Exclude="%(NodeJsExtensionsTargetDirectories.Identity)/**/_temp_/**/*"
```
Comment (0liver) [inline Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets:172]: I've added this to fix an intermittent error that sometimes showed up, and it happened to me again, today:

image

File: Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1 (hunk 2 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/build/Invoke-Sequentially.ps1:40]: You can resolve this conversation then.

File: Lombiq.NodeJs.Extensions/build/NodeJsExtensionsNpmPackage.targets [hunk 4]
```diff
@@ -29,39 +29,16 @@
...
-      <Output TaskParameter="ExitCode" PropertyName="PnpmInstallExitCode" />
-    </Exec>
-    <!-- Use a random back-off <=10s before retrying a failed call to prevent further concurrent calls. -->
-    <PropertyGroup>
-      <TimeoutSeconds>$([MSBuild]::Modulo($([System.DateTime]::Now.Millisecond), 10))</TimeoutSeconds>
-      <Sleep>ping -n $(TimeoutSeconds) 127.0.0.1 >NUL</Sleep>
-    </PropertyGroup>
-    <Exec Command="$(Sleep)" Condition="'$(PnpmInstallExitCode)' != '0'" IgnoreExitCode="true" />
-    <!-- The last try. -->
-    <Exec Command="$(PnpmInstallCommand)" WorkingDirectory="$(NodeJsExtensionsNpmPackageSourcePath)" Condition="'$(PnpmInstallExitCode)' != '0'"/>
```
Comment (0liver) [inline Lombiq.NodeJs.Extensions/build/NodeJsExtensionsNpmPackage.targets:64]: All of this has been replaced by the `MSBuild` target call below 😁

**Review by dministro — APPROVED** [review]

**Pull request merged**